Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,805 @@
+ * writeOrientation also wrote a translation. If true, then no translation was
+ * written to file, and the dependency for the orientation dataset is itself.
+ *
+ * Compliant to the Mantid Instrument Definition file, if a translation exists,
+ * it precedes a rotation.
+ * https://docs.mantidproject.org/nightly/concepts/InstrumentDefinitionFile.html
+ */
+inline void writeOrientation(H5::Group &grp, const Eigen::Quaterniond &rotation,
+                             bool noTranslation) {
+
+  // dependency for orientation defaults to self-dependent. If Location dataset
+  // exists, the orientation will depend on it instead.
+  std::string dependency =
+      !noTranslation ? H5_OBJ_NAME(grp) + "/" + LOCATION : NO_DEPENDENCY;
+
+  double angle;
+
+  H5::DataSet orientation;
+  H5::DataSpace dspace;
+  H5::DataSpace aspace;
+
+  H5::Attribute vector;
+  H5::Attribute units;
+  H5::Attribute transformationType;
+  H5::Attribute dependsOn;
+
+  H5::StrType strSize;
+
+  int rank = 1;
+  hsize_t ddims[static_cast<hsize_t>(1)];
+  ddims[0] = static_cast<hsize_t>(1);
+
+  angle = std::acos(rotation.w()) * (360.0 / PI);
+  Eigen::Vector3d axisOfRotation = rotation.vec().normalized();
+  std::vector<double> stdNormAxis = toStdVector(axisOfRotation);
+
+  dspace = H5Screate_simple(rank, ddims, NULL);
+  orientation =
+      grp.createDataSet(ORIENTATION, H5::PredType::NATIVE_DOUBLE, dspace);
+  orientation.write(&angle, H5::PredType::NATIVE_DOUBLE, dspace);
+
+  int arank = 1;
+  hsize_t adims[static_cast<hsize_t>(3)];
+  adims[0] = static_cast<hsize_t>(3);
+
+  aspace = H5Screate_simple(arank, adims, NULL);
+  vector =
+      orientation.createAttribute(VECTOR, H5::PredType::NATIVE_DOUBLE, aspace);
+  vector.write(H5::PredType::NATIVE_DOUBLE, stdNormAxis.data());
+
+  // units attribute
+  strSize = strTypeOfSize(DEGREES);
+  units = orientation.createAttribute(UNITS, strSize, SCALAR);
+  units.write(strSize, DEGREES);
+
+  // transformation-type attribute
+  strSize = strTypeOfSize(ROTATION);
+  transformationType =
+      orientation.createAttribute(TRANSFORMATION_TYPE, strSize, SCALAR);
+  transformationType.write(strSize, ROTATION);
+
+  // dependency attribute
+  strSize = strTypeOfSize(dependency);
+  dependsOn = orientation.createAttribute(DEPENDS_ON, strSize, SCALAR);
+  dependsOn.write(strSize, dependency);
+}
+
+/*
+ * Function: NXInstrument
+ * for NXentry parent (root group). Produces an NXinstrument group in the parent
+ * group, and writes Nexus compliant datasets and metadata stored in attributes
+ * to the new group.
+ *
+ * @param parent : parent group in which to write the NXinstrument group.
+ * @param compInfo : componentInfo object.
+ */
+H5::Group NXInstrument(const H5::Group &parent,
+                       const Geometry::ComponentInfo &compInfo) {
+
+  std::string nameInCache = compInfo.name(compInfo.root());
+  std::string instrName =
+      nameInCache == "" ? "unspecified_instrument" : nameInCache;
```

Comment:
Prefer to use the `empty()` method when checking for emptiness, on strings or vectors.
Several more cases in this file.

Reply:
done